cli/commands/catalog/module/repo.go

```diff
@@ -0,0 +1,257 @@
+func module(repoName, repoPath, moduleDir string) (*Module, error) {
+	var (
+		modulePath = filepath.Join(repoPath, moduleDir)
+
+		reg       = mdHeaderReg
+		docHeader = mdHeader
+
+		title            = filepath.Base(moduleDir)
+		descriptionLines []string
+	)
+
+	docPath := moduleDocPath(modulePath)
+	if docPath == "" {
+		return nil, nil
+	}
+
+	log.Debugf("Found Terragrunt module in directory %q", modulePath)
+
+	docContentByte, err := os.ReadFile(docPath)
+	if err != nil {
+		return nil, errors.WithStackTrace(err)
+	}
+	docContent := string(docContentByte)
+
+	if strings.HasSuffix(docPath, ".adoc") {
+		reg = adocHeaderReg
+		docHeader = adocHeader
+	}
+
+	if match := reg.FindStringSubmatch(docContent); len(match) > 0 {
+		header := match[1]
+
+		// remove comments
+		header = commentReg.ReplaceAllString(header, "")
+		// remove adoc images
+		header = adocImageReg.ReplaceAllString(header, "")
+
+		lines := strings.Split(header, "\n")
+		title = strings.TrimSpace(lines[0])
+
+		if len(lines) > 1 {
+			for _, line := range lines[1:] {
+				line = strings.TrimSpace(line)
+
+				// another header begins
+				if strings.HasPrefix(line, docHeader) {
+					break
+				}
+
+				descriptionLines = append(descriptionLines, line)
+			}
+		}
+	}
+
+	return &Module{
+		path:        modulePath,
+		url:         path.Join(gitHubOrgURL, repoName, "tree/master", moduleDir),
+		title:       title,
+		description: strings.TrimSpace(strings.Join(descriptionLines, " ")),
+		content:     docContent,
+	}, nil
+
+}
+
+// FindModules clones the repository if `repoPath` is a URL, searches for Terragrunt modules, indexes their README.* files, and returns module instances.
+func FindModules(ctx context.Context, repoPath string) (Modules, error) {
+	var repoName string
+
+	tempDir, err := os.MkdirTemp("", "catalog-*")
+	if err != nil {
+		return nil, errors.WithStackTrace(err)
+	}
+	//nolint:errcheck
+	defer os.RemoveAll(tempDir)
+
+	repoPath, err = getRepo(ctx, repoPath, tempDir)
+	if err != nil {
+		return nil, err
+	}
+
+	remoteURL, err := gitRemoteURL(repoPath)
+	if err != nil {
+		return nil, err
+	}
+	// remove extension like `.git`
+	ext := filepath.Ext(remoteURL)
+	remoteURL = strings.TrimRight(remoteURL, "."+ext)
+
+	repoName = filepath.Base(remoteURL)
+
+	modulesPath := filepath.Join(repoPath, "modules")
+
+	// It is assumed that the repository is a module itself
+	if !files.FileExists(modulesPath) {
+		module, err := module(repoName, repoPath, "")
+		if module == nil || err != nil {
+			return nil, err
+		}
+
+		return Modules{module}, nil
+	}
+
+	var modules Modules
+
+	err = filepath.Walk(modulesPath,
+		func(dir string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if !info.IsDir() {
+				return nil
+			}
+
+			moduleDir, err := filepath.Rel(repoPath, dir)
+			if err != nil {
+				return errors.WithStackTrace(err)
+			}
+
+			module, err := module(repoName, repoPath, moduleDir)
```

Comment:
NIT: variable and function are both called `module`, which is confusing.

Reply:
Rewrote the code.